src/main/scala/akka/stream/contrib/AccumulateWhileUnchanged.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.contrib
+
+import akka.japi.function
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler }
+import akka.stream.{ Attributes, FlowShape, Inlet, Outlet }
+
+import scala.collection.immutable
+
+object AccumulateWhileUnchanged {
+
+  /**
+   * Factory for [[AccumulateWhileUnchanged]] instances
+   *
+   * @param propertyExtractor a function to extract the observed element property
+   * @tparam Element  type of accumulated elements
+   * @tparam Property type of the observed property
+   * @return [[AccumulateWhileUnchanged]] instance
+   */
+  def apply[Element, Property](propertyExtractor: Element => Property) = new AccumulateWhileUnchanged(propertyExtractor)
+
+  /**
+   * Java API: Factory for [[AccumulateWhileUnchanged]] instances
+   *
+   * @param propertyExtractor a function to extract the observed element property
+   * @tparam Element  type of accumulated elements
+   * @tparam Property type of the observed property
+   * @return [[AccumulateWhileUnchanged]] instance
+   */
+  def create[Element, Property](propertyExtractor: function.Function[Element, Property]) = new AccumulateWhileUnchanged(propertyExtractor.apply)
+}
+
+/**
+ * Accumulates elements of type [[Element]] while a property extracted with [[propertyExtractor]] remains unchanged,
+ * emits an accumulated sequence when the property changes
+ *
+ * @param propertyExtractor a function to extract the observed element property
+ * @tparam Element  type of accumulated elements
+ * @tparam Property type of the observed property
+ */
+final class AccumulateWhileUnchanged[Element, Property](propertyExtractor: Element => Property) extends GraphStage[FlowShape[Element, immutable.Seq[Element]]] {
+
+  val in = Inlet[Element]("AccumulateWhileUnchanged.in")
+  val out = Outlet[immutable.Seq[Element]]("AccumulateWhileUnchanged.out")
+
+  override def shape = FlowShape.of(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) {
+
+    private var currentState: Option[Property] = None
+    private val buffer = Vector.newBuilder[Element]
+
+    setHandlers(in, out, new InHandler with OutHandler {
+
+      override def onPush(): Unit = {
+        val nextElement = grab(in)
+        val nextState = propertyExtractor(nextElement)
+
+        if (currentState.isEmpty || currentState.contains(nextState)) {
+          buffer += nextElement
+          pull(in)
+        } else {
+          val result = buffer.result()
+          buffer.clear()
+          buffer += nextElement
+          push(out, result)
+        }
+
+        currentState = Some(nextState)
```

Comment:
case Some(s) if s == nextState =>

==

case Some(`nextState`) =>

-- 
Cheers,
√

On Jul 21, 2016 11:43 AM, "Jacek Kunicki" <notifications@github.com> wrote:

> In src/main/scala/akka/stream/contrib/AccumulateWhileUnchanged.scala
> <https://github.com/akka/akka-stream-contrib/pull/28#discussion_r71676280>
> :
>
> > +
> > +      override def onPush(): Unit = {
> > +        val nextElement = grab(in)
> > +        val nextState = propertyExtractor(nextElement)
> > +
> > +        if (currentState.isEmpty || currentState.contains(nextState)) {
> > +          buffer += nextElement
> > +          pull(in)
> > +        } else {
> > +          val result = buffer.result()
> > +          buffer.clear()
> > +          buffer += nextElement
> > +          push(out, result)
> > +        }
> > +
> > +        currentState = Some(nextState)
>
> With the pattern matching approach, some common logic would be required
> for the None and Some(s) parts (which could of course be extracted to a
> method), so we'd end up with something like:
>
> def appendAndPull(): Unit = {
>   buffer += nextElement
>   pull(in)
> }
>
> currentState match {
>   case None =>
>     currentState = Some(nextState)
>     appendAndPull()
>   case Some(s) if s == nextState =>
>     appendAndPull()
>   case _ =>
>     val result = buffer.result()
>     buffer.clear()
>     buffer += nextElement
>     push(out, result)
>     currentState = Some(nextState)
> }
>
> which is not necessarily better in my opinion (but indeed does less option
> wrapping).
>
> As for the currentState = currentState.orElse(Some(nextState)) - this is
> nothing more than an additional if, just hidden.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/akka/akka-stream-contrib/pull/28/files/72c69890a8d59d016299f99f873232d955b2562b#r71676280>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAqdzQoHCaMSjhzDX-LoPmyvW5catzsks5qXz9cgaJpZM4JF8HB>
> .
>